roc.go

```diff
@@ -0,0 +1,136 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true
+// positive rate (TPR) values corresponding to n
+// cutoffs spanning the relative (or receiver)
+// operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+// input y must be sorted, and SortWeightedLabelled can be used
+// in order to sort data y together with classes and weights.
+// For a given cutoff value, observations corresponding to
+// entries in y greater than the cutoff are classified as
+// false, and these assigned class labels are compared to
+// classes to calculate FPR and TPR.
+// weights being nil results in the unweighted case.
+// n of zero results in all possible cutoffs being calculated.
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.IsSorted(sort.Float64Slice(y)) {
+		panic("stat: input must be sorted")
+	}
+
+	var iw, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot less than 2 points on a ROC curve")
+		}
+		iw = (y[len(y)-1] - y[0]) / float64(n-1)
+		if iw <= 0 {
+			panic("stat: y constant or not sorted")
+		}
+		tol = y[0] + iw
+	}
+
+	if n > (len(y) + 1) {
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	newEntry := true
+	var j int
+	var nPositive, nNegative float64
+	for i, u := range classes {
```

Comment:
I've made changes according to your dot points 1, 3, and 4.

Your second dot-point: newEntry became neccessary when we wanted to include arbitrary number of cutoffs and removal of duplicate values. Essentially it is because the first time we move to a new bin we need to copy the previous values, but if we havent just moved to a new bin then we just need to add to the current values and not copy from the previous values again. One possible solution would be to iterate i within the loop to skip the newEntry==true cases and do them manually when they are detected with 
`if y[i] != y[i+1]`
or
`if y[i+1] > tol . `
I'm not sure if this solution would be cleaner or not though, I like that the loop
`for i, u := range classes`
loop actually goes through len(y) loops and this would change that. 

Your last dot point, as I mentioned above n==0 is not the same as n=len(y)+1, 
n==len(y)+1 will produce FPR and TPR values for len(y)+1 equally spaced cutoffs
and this is not the same thing as producing FPR and TPR values for all possible
cutoffs (the behaviour when n==0). If the values in y are equally spaced, then these
two things are equivalent but otherwise they cannot be guaranteed to be the same, 
so I have deliberately left them as seperate cases.